lib/facterdb.rb

```diff
@@ -123,8 +130,16 @@ def self.valid_filters?(filters)
 
   # @return [Array] - array of hashes of facts
   # @param filter [Object] - The filter to convert to jgrep string
-  def self.get_facts(filter=nil)
+  def self.get_facts(filter=nil, cache=true)
+    if cache && filter && filter == Thread.current[:facterdb_last_filter_seen]
+      return Thread.current[:facterdb_last_facts_seen]
+    end
     filter_str = generate_filter_str(filter)
-    JGrep.jgrep(database, filter_str).map { |hash| Hash[hash.map{ |k, v| [k.to_sym, v] }] }
+    result = JGrep.jgrep(database, filter_str).map { |hash| Hash[hash.map{ |k, v| [k.to_sym, v] }] }
```

Comment:
Note to self: this is why only the keys are symbols while it should really be something like `deep_symbolize` or use something like https://api.rubyonrails.org/classes/ActiveSupport/HashWithIndifferentAccess.html

Reply:
I do not understand the implementation enough, but my general rule of thumb is - for big dictionaries, stay away from symbolizing/indifferent access it can be quite slow. I suggest to use just strings and convert all input to strings as well if possible.

Reply:
Yes, but that's an API change that's quite big. That's why HashWithIndifferentAccess is probably a cleaner solution.